plugin/src/main/java/org/opensearch/ml/action/models/UpdateModelTransportAction.java

```diff
@@ -0,0 +1,326 @@
+import org.opensearch.ml.model.MLModelManager;
+import org.opensearch.ml.utils.MLNodeUtils;
+import org.opensearch.ml.utils.RestActionUtils;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.builder.SearchSourceBuilder;
+import org.opensearch.search.fetch.subphase.FetchSourceContext;
+import org.opensearch.tasks.Task;
+import org.opensearch.transport.TransportService;
+
+import lombok.AccessLevel;
+import lombok.experimental.FieldDefaults;
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2
+@FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
+public class UpdateModelTransportAction extends HandledTransportAction<ActionRequest, UpdateResponse> {
+    Client client;
+    NamedXContentRegistry xContentRegistry;
+    ModelAccessControlHelper modelAccessControlHelper;
+    ConnectorAccessControlHelper connectorAccessControlHelper;
+    MLModelManager mlModelManager;
+
+    @Inject
+    public UpdateModelTransportAction(
+        TransportService transportService,
+        ActionFilters actionFilters,
+        Client client,
+        NamedXContentRegistry xContentRegistry,
+        ConnectorAccessControlHelper connectorAccessControlHelper,
+        ModelAccessControlHelper modelAccessControlHelper,
+        MLModelManager mlModelManager
+    ) {
+        super(MLUpdateModelAction.NAME, transportService, actionFilters, MLUpdateModelRequest::new);
+        this.client = client;
+        this.xContentRegistry = xContentRegistry;
+        this.modelAccessControlHelper = modelAccessControlHelper;
+        this.connectorAccessControlHelper = connectorAccessControlHelper;
+        this.mlModelManager = mlModelManager;
+    }
+
+    @Override
+    protected void doExecute(Task task, ActionRequest request, ActionListener<UpdateResponse> actionListener) {
+        MLUpdateModelRequest updateModelRequest = MLUpdateModelRequest.fromActionRequest(request);
+        MLUpdateModelInput updateModelInput = updateModelRequest.getUpdateModelInput();
+        String modelId = updateModelInput.getModelId();
+        UpdateRequest updateRequest = new UpdateRequest(ML_MODEL_INDEX, modelId);
+
+        try {
+            updateRequest.doc(updateModelInput.toXContent(XContentFactory.jsonBuilder(), ToXContent.EMPTY_PARAMS));
+        } catch (IOException e) {
+            log.error("Failed to construct the update request with the input.", e);
+            actionListener.onFailure(e);
+        }
+        updateRequest.docAsUpsert(true);
+        User user = RestActionUtils.getUserContext(client);
+
+        MLModelGetRequest mlModelGetRequest = new MLModelGetRequest(modelId, false);
+        FetchSourceContext fetchSourceContext = getFetchSourceContext(mlModelGetRequest.isReturnContent());
+        GetRequest getModelRequest = new GetRequest(ML_MODEL_INDEX).id(modelId).fetchSourceContext(fetchSourceContext);
+
+        try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
```

Comment:
You stash context here, but seems not restore context

Reply:
So do we need a `ActionListener.runBefore`?